src/main/java/data_access/UserDataAccessObject.java

```diff
@@ -19,70 +18,98 @@
+    }
+
+    @Override
+    public boolean existsByNameStudent(String username) {
+        boolean found = false;
+        for (Student student : studentArrayList) {
+            if (student.getUsername().equals(username)) {
+                found = true;
+                break;
             }
         }
         return found;
     }
 
     @Override
-    public boolean existsByEmail(String identifier) {
+    public boolean existsByEmailClub(String identifier) {
         boolean found = false;
-        for (User user: userArrayList) {
-            if (user.getEmail().equals(identifier)) {
+        for (Club club : clubArrayList) {
+            if (club.getEmail().equals(identifier)) {
                 found = true;
+                break;
+            }
+        }
+        return found;
+    }
+
+    @Override
+    public boolean existsByEmailStudent(String identifier) {
+        boolean found = false;
+        if (!found) {
+            for (Student student : studentArrayList) {
+                if (student.getEmail().equals(identifier)) {
+                    found = true;
+                    break;
+                }
             }
         }
         return found;
     }
 
     @Override
-    public void saveClub(User club) {
-        userArrayList.add((Club) club);
+    public void saveClub(Club club) {
+        clubArrayList.add(club);
     }
 
     @Override
-    public void saveStudent(User student) {
-        userArrayList.add((Student) student);
+    public void saveStudent(Student student) {
+        studentArrayList.add(student);
     }
 
     @Override
     public Club getClub(String email) {
-        Club foundClub = null;
-        for (User user : userArrayList) {
-            if (user.getEmail().equals(email)) {
-                foundClub = (Club) user;
+        Club clubFound = null;
+        for (Club club : clubArrayList) {
+            if (club.getEmail().equals(email)) {
+                clubFound = club;
             }
         }
-        return foundClub;
+        // This should not be returned as null since the precondition states that the club must exist.
+        return clubFound;
     }
 
     @Override
     public Student getStudent(String email) {
         Student foundStudent = null;
-        for (User user : userArrayList) {
-            if (user.getEmail().equals(email)) {
-                foundStudent = (Student) user;
+        for (Student student : studentArrayList) {
+            if (student.getEmail().equals(email)) {
+                foundStudent = student;
             }
         }
+        // This should not be returned as null since the precondition states that the student must exist.
         return foundStudent;
     }
 
     @Override
     public void savePost(Post post, Club club) {
-        for (User user : userArrayList) {
-            if (user.getEmail().equals(club.getEmail())) {
-                final Club isClub = (Club) user;
-                isClub.addClubPost(post);
-                break;
-
```

Comment:
Sorry if this is a stupid question but is there a better way to do this than to loop through a whole list? What happens when we implement the firebase firestore?

Reply:
You are 100% right. It will be faster to implement it using an email search instead for the firebase DAO. However, with the way data is stored currently, it does not matter if I look at the name or the email as the run time would be the exact same (We have an ArrayList of Clubs, not a map with key email).
However, I can always add a break to stop the loop early.

Reply:
awesome thanks this makes perfect sense